src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -738,6 +743,23 @@ private Exception WrapException(Exception ex, string action)
             }
         }
 
+        private Exception WrapInOperationCanceledExceptionIfRequired(Exception exception, CancellationToken cancellationToken)
+        {
+            if (exception is ObjectDisposedException objectDisposedException
+                &&
+                (
+                    objectDisposedException.ObjectName == GetType().Name ||
+                    objectDisposedException.Message == "The semaphore has been disposed."
+                ))
+            {
+                return new OperationCanceledException($"The {nameof(BinaryConnection)} has been cancelled.", exception);
```

Comment:
I don't think the message is quite right. "Classes" don't get cancelled, "operations" do. Perhaps:

    return new OperationCanceledException($"The {nameof(BinaryConnection)} operation has been cancelled.", exception);

Reply:
done